Ok, importing modules and copying the funciton we made in 01_beautifulsoup_setup.ipynb

In [1]:
import requests
from bs4 import BeautifulSoup
from time import sleep
import pandas as pd

EDIT: After finishing and testing, I added the for loop to check for the keywords before continuing.

In [2]:
keywords = ['mattress', 'cabinet', 'wrench']
def grabpost(thisurl):
    thispage = requests.get(thisurl)
    bs = BeautifulSoup(thispage.text,'html.parser')
    post_body = bs.find(id='postingbody').text
    
    for word in keywords:
        if word in post_body:
        title = bs.find(id='titletextonly').text
        time_posted = bs.find('time').text.strip()
        address = bs.find(class_='mapaddress').text
        
        return {'url':thisurl, 'title': title, 'time': time_posted,
               'address': address, 'post_body': post_body}

Just gonna make a quick csv with the headders we're using in the dictionary created by grabpost() before we move on.

In [3]:
filename = 'craigslist.csv'
with open(filename,'w') as f:
  f.write('url,title,time,address,post_body\n')

In [18]:
df = pd.read_csv(filename)

ok, grabbing our html parser from bs4 to get try and find the links for all the posts on the Phoenix Garage & Moving Sales page.

In [4]:
thispage = requests.get('https://phoenix.craigslist.org/search/gms')
bs = BeautifulSoup(thispage.text,'html.parser')

In [15]:
links = bs.find_all(class_='result-title hdrlnk')
links

[<a class="result-title hdrlnk" data-id="7442684354" href="https://phoenix.craigslist.org/nph/gms/d/phoenix-reduced-kids-motorized-frozen/7442684354.html" id="postid_7442684354">REDUCED Kids Motorized Frozen Benz, BMW i8 Coupe, 4 Wheeler....</a>,
 <a class="result-title hdrlnk" data-id="7439457025" href="https://phoenix.craigslist.org/nph/gms/d/phoenix-chevy-wheels-lb-10-lb-weights/7439457025.html" id="postid_7439457025">Chevy Wheels  5 lb 10 lb weights and heavy bag speed bag stand$$</a>,
 <a class="result-title hdrlnk" data-id="7444018225" href="https://phoenix.craigslist.org/nph/gms/d/phoenix-garage-sale-thursday-friday-feb/7444018225.html" id="postid_7444018225">Garage Sale Thursday &amp; Friday Feb 10,11</a>,
 <a class="result-title hdrlnk" data-id="7443953320" href="https://phoenix.craigslist.org/wvl/gms/d/sun-city-west-huge-yard-and-plant-sale/7443953320.html" id="postid_7443953320">Huge Yard and Plant SALE</a>,
 <a class="result-title hdrlnk" data-id="7444034924" href="https://

In [16]:
links[1]['href']

'https://phoenix.craigslist.org/nph/gms/d/phoenix-chevy-wheels-lb-10-lb-weights/7439457025.html'

Got it! Now to make our for loop.

In [19]:
for eachlink in links:
    full_url = eachlink['href']
    if not df.url.str.contains(full_url).any():
        print('getting',full_url)
        post = grabpost(full_url)
        sleep(15)
        df = df.append(post,ignore_index=True)

getting https://phoenix.craigslist.org/nph/gms/d/phoenix-reduced-kids-motorized-frozen/7442684354.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/nph/gms/d/phoenix-chevy-wheels-lb-10-lb-weights/7439457025.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/nph/gms/d/phoenix-garage-sale-thursday-friday-feb/7444018225.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/sun-city-west-huge-yard-and-plant-sale/7443953320.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/nph/gms/d/phoenix-huge-estate-sale/7444034924.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/sun-city-estate-sale/7442666811.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/palo-verde-huge-yard-sale/7444063257.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/evl/gms/d/rio-verde-yard-estate-sale/7444089826.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/sun-city-west-patio-sale/7443808092.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/sun-city-west-moving-sale/7444046776.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/glendale-big-garage-sale/7443928236.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/evl/gms/d/mesa-multi-family-garage-sale-thursday/7444053100.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/nph/gms/d/glendale-estate-salegarage-saleyard/7441028255.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/peoria-moving-sale/7441914080.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/evl/gms/d/gold-canyon-moving-sale/7442605011.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/sun-city-west-moving-sale/7443649101.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/peoria-large-moving-sale/7443658618.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/evl/gms/d/tempe-moving-sale-many-items-come-get/7443665673.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/cph/gms/d/phoenix-everything-must-go/7439354831.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/nph/gms/d/phoenix-speed-bag-stand35-four-wheels/7439591056.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/phoenix-carport-sale/7443528430.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/nph/gms/d/phoenix-yard-sale-yard-sale/7443723006.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/evl/gms/d/mesa-patio-sales/7444093053.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/wvl/gms/d/luke-air-force-base-garage-sale/7444018876.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/evl/gms/d/scottsdale-estate-garage-sale/7443923617.html


/tmp/ipykernel_4746/609372375.py:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(post,ignore_index=True)


getting https://phoenix.craigslist.org/evl/gms/d/mesa-yard-sale/7443979379.html


KeyboardInterrupt: 

In [21]:
df

url  \
0   https://phoenix.craigslist.org/nph/gms/d/phoen...   
1   https://phoenix.craigslist.org/nph/gms/d/phoen...   
2   https://phoenix.craigslist.org/nph/gms/d/phoen...   
3   https://phoenix.craigslist.org/wvl/gms/d/sun-c...   
4   https://phoenix.craigslist.org/nph/gms/d/phoen...   
5   https://phoenix.craigslist.org/wvl/gms/d/sun-c...   
6   https://phoenix.craigslist.org/wvl/gms/d/palo-...   
7   https://phoenix.craigslist.org/evl/gms/d/rio-v...   
8   https://phoenix.craigslist.org/wvl/gms/d/sun-c...   
9   https://phoenix.craigslist.org/wvl/gms/d/sun-c...   
10  https://phoenix.craigslist.org/wvl/gms/d/glend...   
11  https://phoenix.craigslist.org/evl/gms/d/mesa-...   
12  https://phoenix.craigslist.org/nph/gms/d/glend...   
13  https://phoenix.craigslist.org/wvl/gms/d/peori...   
14  https://phoenix.craigslist.org/evl/gms/d/gold-...   
15  https://phoenix.craigslist.org/wvl/gms/d/sun-c...   
16  https://phoenix.craigslist.org/wvl/gms/d/peori...   
17  https://phoenix.craigslist.org/evl/gms/d/tempe...   
18  https://phoenix.craigslist.org/cph/gms/d/phoen...   
19  https://phoenix.craigslist.org/nph/gms/d/phoen...   
20  https://phoenix.craigslist.org/wvl/gms/d/phoen...   
21  https://phoenix.craigslist.org/nph/gms/d/phoen...   
22  https://phoenix.craigslist.org/evl/gms/d/mesa-...   
23  https://phoenix.craigslist.org/wvl/gms/d/luke-...   
24  https://phoenix.craigslist.org/evl/gms/d/scott...   

                                                title              time  \
0   REDUCED Kids Motorized Frozen Benz, BMW i8 Cou...  2022-02-06 17:44   
1   Chevy Wheels  5 lb 10 lb weights and heavy bag...  2022-01-30 03:35   
2             Garage Sale Thursday & Friday Feb 10,11  2022-02-09 17:18   
3                            Huge Yard and Plant SALE  2022-02-09 14:40   
4                       HUGE estate sale 2/10, 11, 12  2022-02-09 18:07   
5                                         Estate Sale  2022-02-06 16:48   
6                                      huge yard sale  2022-02-09 19:46   
7                                    Yard/estate sale  2022-02-09 21:58   
8                                          Patio Sale  2022-02-09 10:11   
9                                         Moving sale  2022-02-09 18:45   
10                                 Big Garage Sale!!!  2022-02-09 13:49   
11       multi family garage sale--thursday 2/10  7am  2022-02-09 19:06   
12  ~ESTATE SALE!~GARAGE SALE~YARD SALE~Everything...  2022-02-02 17:52   
13                                        MOVING SALE  2022-02-04 19:02   
14                                        Moving Sale  2022-02-06 14:05   
15                                      Moving Sale -  2022-02-08 22:20   
16                                 Large moving sale!  2022-02-08 23:33   
17            MOVING SALE: MANY ITEMS - COME GET THEM  2022-02-09 00:54   
18                                 Everything Must Go  2022-01-29 16:41   
19  speed bag stand$35. Four wheels $150 Gold gym ...  2022-01-30 11:36   
20                                       CARPORT SALE  2022-02-08 15:25   
21                             Yard Sale!  Yard Sale!  2022-02-09 07:30   
22                                        Patio sales  2022-02-09 22:19   
23                                        Garage Sale  2022-02-09 17:20   
24                                Estate Garage Sale:  2022-02-09 13:40   

                                             address  \
0                                        Central Ave   
1              15th,ave near 15th avenue and Hatcher   
2             julie Dr near unionhills Dr & 31st ave   
3                                    15629 N 99th Dr   
4                                3302 W. Pontiac Dr.   
5                                99th Ave near Grand   
6                                      201 7th Ave E   
7                               156th near Rio Verde   
8                     13087 N 100 Ave near Forrester   
9                       21628 N 155 Drive near 151st   


In [20]:
len(links)

188

Looks like it's working, so I intturpted the kernel. We got 188 links, so at 15 sec each, should take about 45min-1hr. I'll go ahead and put it all in a .py file and run it.... oh, wait, forgot to add keywords and actually write to the csv. I'll put they keyword part in the function we created earlier.